Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ProjectionContentGraph.php

```diff
@@ -262,7 +262,15 @@ public function determineHierarchyRelationPosition(
                 ]
             )->fetchAssociative();
 
+            if (!$succeedingSiblingRelation) {
+                throw new \Exception(
```

Comment:
Do we throw the base exception in similar cases or is there a more specific exception?

Reply:
Usually we just accept the null and handle it by returning null or zero in methods. I found one occurrence where we throw a runtime exception when we cannot fetch a node: throw new \RuntimeException(sprintf('Failed to fetch node: %s', $e->getMessage()), 1678286030, $e); 

I would suggest using something like a UnexpectedValueException here, since we kind of do expect the succeedingSibling to exist if it is passed to the method.

Reply:
Sorry, I didn't realize that this code is invoked from a _projection_.. Throwing exceptions in projections is the last resort as there is no easy way out..
We should find the root cause of the error, i.e.

> there is something wrong with the NodeRelationAnchor point

and check if a bug led to this state.

Maybe you can trace back the code to find out which event triggered the error?

Reply:
The error occurred only once during one of many migration attempts and I cannot definitely say how I got there. There were lots of errors during lots of attempts, and I can only assume that  it has something to do with this bug: https://github.com/neos/neos-development-collection/issues/4504#issuecomment-1735759814 

Showing an error is an improvement over just having an offset error shown, as it is happening right now. It is worth having a discussion what the expectation would be if the combination of contentstreamID, succeedingSiblingAnchorPoint and dimensionspacepoint does not return a viable relation. Do we use the parentAnchorPoint as a "fallback" or do we assume that at this point something went horribly wrong and show an error?

Reply:
Thanks for the clarification!

> Showing an error is an improvement over just having an offset error shown

Absolutely! My point is, that it is too late at this point. The event has already been published with (potentially) invalid data so we should try to detect and prevent this case _earlier_ already.

> It is worth having a discussion what the expectation would be if the combination of contentstreamID, succeedingSiblingAnchorPoint and dimensionspacepoint does not return a viable relation

Yes!

Reply:
Maybe this bug is solve with all the adjustments from @nezaniel lately?

Reply:
As far as I can see this is a bug in the migration script and cannot be solved here

Reply:
Ill merge this as indeed we throw in many other places also `RuntimeException`'s in the projection even if the case can never happen. Its better than to obscure it. Phpstan should actually enforce this :D